### PR TITLE
Fixed #941.

### LINKED ISSUE #941: Possible race conditions with `MappingFEField`
PR #935 addressed a race condition issue with a possibly not definitive fix. Local data should be moved inside InternalData.

### PR DESCRIPTION
**Pull request by luca-heltai**
All thread sensible data was moved from MappingFEField inside MappingFEField::InternalData, which is constructed locally and therefore is intrinsically threadsafe.

All tests pass, except the ones addressed by #1195.

### CODE REVIEW

**Comment by luca-heltai** [comment]
The two new fields should be  `mutable` once #1182 is merged.

File: source/fe/mapping_fe_field.cc [hunk 1]
```diff
@@ -784,13 +779,13 @@ transform_unit_to_real_cell (const typename Triangulation<dim,spacedim>::cell_it
 //  Use the get_data function to create an InternalData with data vectors of
 //  the right size and transformation shape values already computed at point
 //  p.
-  update_internal_dofs(cell);
-
   const Quadrature<dim> point_quadrature(p);
   std_cxx11::unique_ptr<InternalData>
   mdata (dynamic_cast<InternalData *> (
            get_data(update_transformation_values, point_quadrature)));
```
Comment (kronbichler) [inline source/fe/mapping_fe_field.cc]: Even though not part of the patch: You can get rid of the dynamic_cast here by changing the return value of get_data to MappingFEField::InternalData, see #1183.

**Comment by kronbichler** [comment]
This looks very good and does exactly what was proposed in #941.

**Comment by luca-heltai** [comment]
@kronbichler : thanks for the suggestion! I didn't know you could do that... :+1:

**Comment by kronbichler** [comment]
Nice.

**Comment by kronbichler** [comment]
Candidate for 8.3 branch?

**Comment by luca-heltai** [comment]
yup. I'd put it in together with #1195.

@tjhei ? 

#1195

**Comment by bangerth** [comment]
Great, I think that's exactly the right way to fix this problem!
